### PR TITLE
FT: Add PartNumber support for getObject

### PR DESCRIPTION
**Pull request by bennettbuchanan**
Depends on https://github.com/scality/Arsenal/pull/309.

### CODE REVIEW

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: lib/api/objectGet.js [hunk 1]
```diff
@@ -106,6 +106,23 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
             // objMD.location is just a string
             dataLocator = Array.isArray(objMD.location) ?
                 objMD.location : [{ key: objMD.location }];
+
+            let partNumber = null;
+            if (Object.keys(request.query).includes('partNumber')) {
```
Comment (nicolas2bert) [inline lib/api/objectGet.js]: Could you also do: if(request.query.partNumber !== undefined)?

File: lib/api/objectGet.js [hunk 2]
```diff
@@ -106,6 +106,23 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
...
+                if (byteRange) {
+                    const error = errors.InvalidRequest
```
Comment (nicolas2bert) [inline lib/api/objectGet.js]: we might want to log here. :)

File: lib/api/objectGet.js [hunk 3]
```diff
@@ -106,6 +106,23 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
...
+                        .customizeDescription('Cannot specify both Range ' +
+                            'header and partNumber query parameter.');
+                    return callback(error, null, corsHeaders);
+                }
+                partNumber = request.query.partNumber;
+                if (partNumber < 1 || partNumber > 10000) {
+                    const error = errors.InvalidArgument
```
Comment (nicolas2bert) [inline lib/api/objectGet.js]: log here too?

File: lib/api/objectGet.js [hunk 4]
```diff
@@ -120,7 +137,18 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
                         objMD['x-amz-server-side-encryption'];
                 }
             }
-            dataLocator = setPartRanges(dataLocator, byteRange);
+            if (partNumber) {
+                const partObj = dataLocator[partNumber - 1];
+                if (partObj === undefined) {
+                    return callback(errors.InvalidPartNumber, null,
```
Comment (nicolas2bert) [inline lib/api/objectGet.js]: log

File: tests/functional/aws-node-sdk/test/object/get.js [hunk 5]
```diff
@@ -50,6 +52,83 @@ describe('GET object', () => {
...
+
+            return async.waterfall([
+                next => {
+                    const createMpuParams = {
+                        Bucket: bucketName,
+                        Key: objectName,
+                    };
+
+                    s3.createMultipartUpload(createMpuParams, (err, data) => {
+                        if (err) {
```
Comment (nicolas2bert) [inline tests/functional/aws-node-sdk/test/object/get.js]: You could have used your cool `checkNoError` here and after :)

File: lib/api/objectGet.js (hunk 2 above)
Comment (ploki) [inline lib/api/objectGet.js]: These are client 4XX errors, we're better not logging these

File: lib/api/objectGet.js (hunk 1 above)
Comment (bennettbuchanan) [inline lib/api/objectGet.js]: Updated.

File: tests/functional/aws-node-sdk/test/object/get.js (hunk 5 above)
Comment (bennettbuchanan) [inline tests/functional/aws-node-sdk/test/object/get.js]: Updated. Not mine, but it is a cool one! 😸

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ploki I have run these tests successfully with `AWS_ON_AIR=true`.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by ploki** [comment]
Great! thanks!

File: lib/api/objectGet.js [hunk 6]
```diff
@@ -106,6 +106,28 @@ function objectGet(authInfo, request, returnTagCount, log, callback) {
...
+            let partNumber = null;
+            if (request.query && request.query.partNumber !== undefined) {
+                if (byteRange) {
+                    const error = errors.InvalidRequest
+                        .customizeDescription('Cannot specify both Range ' +
+                            'header and partNumber query parameter.');
+                    return callback(error, null, corsHeaders);
+                }
+                partNumber = Number(request.query.partNumber, 10);
+                if (Number.isNaN(partNumber)) {
```
Comment (bennettbuchanan) [inline lib/api/objectGet.js]: I have added this check, even though the AWS SDK checks that the `PartNumber` param is a `Number`.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Pull request merged**